x/feemarket/ante/fee.go

```diff
@@ -41,96 +45,109 @@ func (dfd FeeMarketCheckDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simula
+		if feeDenom != requiredBaseFee.Denom {
+			return ctx, err
+		}
+	}
+
 	ctx.Logger().Info("fee deduct ante handle",
-		"min gas prices", minGasPrices,
-		"fee", fee,
+		"min gas prices", requiredBaseFee,
+		"fee", feeCoins,
 		"gas limit", gas,
 	)
 
 	if !simulate {
-		fee, _, err = CheckTxFees(ctx, minGasPrices, feeTx, true)
+		feeCoin, _, err = CheckTxFee(ctx, requiredBaseFee, feeTx, true, dfd.feemarketKeeper.GetDenomResolver())
 		if err != nil {
 			return ctx, errorsmod.Wrapf(err, "error checking fee")
 		}
 	}
 
-	minGasPricesDecCoins := sdk.NewDecCoinsFromCoins(minGasPrices...)
-	newCtx := ctx.WithPriority(getTxPriority(fee, int64(gas))).WithMinGasPrices(minGasPricesDecCoins)
+	minGasPricesDecCoin := sdk.NewDecCoinFromCoin(requiredBaseFee)
+	newCtx := ctx.WithPriority(getTxPriority(feeCoin, int64(gas))).WithMinGasPrices(sdk.NewDecCoins(minGasPricesDecCoin))
 	return next(newCtx, tx, simulate)
 }
 
-// CheckTxFees implements the logic for the fee market to check if a Tx has provided sufficient
+// CheckTxFee implements the logic for the fee market to check if a Tx has provided sufficient
 // fees given the current state of the fee market. Returns an error if insufficient fees.
-func CheckTxFees(ctx sdk.Context, minFees sdk.Coins, feeTx sdk.FeeTx, isCheck bool) (feeCoins sdk.Coins, tip sdk.Coins, err error) {
-	minFeesDecCoins := sdk.NewDecCoinsFromCoins(minFees...)
-	feeCoins = feeTx.GetFee()
+func CheckTxFee(ctx sdk.Context, minFee sdk.Coin, feeTx sdk.FeeTx, isCheck bool, resolver feemarkettypes.DenomResolver) (feeCoin sdk.Coin, tip sdk.Coin, err error) {
+	minFeesDecCoin := sdk.NewDecCoinFromCoin(minFee)
+	feeCoin = feeTx.GetFee()[0]
+
+	feeCoin, err = resolver.ConvertToBaseToken(ctx, feeCoin, minFee.Denom)
+	if err != nil {
+		return sdk.Coin{}, sdk.Coin{}, err
+	}
 
 	// Ensure that the provided fees meet the minimum
-	minGasPrices := minFeesDecCoins
-	if !minGasPrices.IsZero() {
-		requiredFees := make(sdk.Coins, len(minGasPrices))
-		consumedFees := make(sdk.Coins, len(minGasPrices))
+	minGasPrice := minFeesDecCoin
+	if !minGasPrice.IsZero() {
+		var requiredFee sdk.Coin
+		var consumedFee sdk.Coin
 
 		// Determine the required fees by multiplying each required minimum gas
 		// price by the gas, where fee = ceil(minGasPrice * gas).
 		gasConsumed := int64(ctx.GasMeter().GasConsumed())
 		gcDec := sdkmath.LegacyNewDec(gasConsumed)
 		glDec := sdkmath.LegacyNewDec(int64(feeTx.GetGas()))
 
-		for i, gp := range minGasPrices {
-			consumedFee := gp.Amount.Mul(gcDec)
-			limitFee := gp.Amount.Mul(glDec)
-			consumedFees[i] = sdk.NewCoin(gp.Denom, consumedFee.Ceil().RoundInt())
-			requiredFees[i] = sdk.NewCoin(gp.Denom, limitFee.Ceil().RoundInt())
-		}
-
-		if !feeCoins.IsAnyGTE(requiredFees) {
-			return nil, nil, sdkerrors.ErrInsufficientFee.Wrapf(
-				"got: %s required: %s, minGasPrices: %s, gas: %d",
-				feeCoins,
-				requiredFees,
-				minGasPrices,
+		consumedFeeAmount := minGasPrice.Amount.Mul(gcDec)
```

Comment:
this is the gas utilization prior to the transaction being executed i.e. the messages?

Reply:
Yes we don't use this in the `ante` check. Only in the `post` check when gas has actually been consumed